src/dodona/junit/MultiMessageWriter.java

```diff
@@ -0,0 +1,118 @@
+package dodona.junit;
+
+import dodona.feedback.Format;
+import dodona.feedback.Message;
+import dodona.feedback.Permission;
+import org.junit.rules.TestRule;
+import org.junit.runner.Description;
+import org.junit.runners.model.Statement;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.function.Predicate;
+
+/**
+ * Feedback message writer that allows multiple messages and rewinding of
+ * previously sent messages.
+ */
+public class MultiMessageWriter implements TestRule {
+    private final Format defaultFormat;
+    private final Permission defaultPermission;
+    
+    private final List<Message> messages;
+    
+    /**
+     * MultiMessageWriter constructor.
+     *
+     * @param defaultPermission the default permission for messages
+     * @param defaultFormat     the default format for messages
+     */
+    public MultiMessageWriter(final Permission defaultPermission, final Format defaultFormat) {
+        this.defaultFormat = defaultFormat;
+        this.defaultPermission = defaultPermission;
+        this.messages = new ArrayList<>();
+    }
+    
+    /**
+     * MultiMessageWriter constructor. Assumes student-level code messages.
+     */
+    public MultiMessageWriter() {
+        this(Permission.STUDENT, Format.CODE);
+    }
+    
+    @Override
+    public Statement apply(final Statement base, final Description description) {
+        return new Statement() {
+            @Override
+            public void evaluate() throws Throwable {
+                try {
+                    base.evaluate();
+                } catch (final TestCarryingThrowable e) {
+                    throw new TestCarryingThrowable(e.getCause(), e.getStartTest(), MultiMessageWriter.this.getMessages(), e.getCloseTest());
```

Comment:
We'd require less code changes if you'd utilize the recursive character of the Throwables here. Since we loop over the causes to take their messages anyway...